File: src/contact_pkg/contact_pkg/feasibility_test_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
# [수정 1] WrenchStamped 메시지 타입 임포트 추가
from geometry_msgs.msg import WrenchStamped

class FeasibilityTestNode(Node):
    def __init__(self):
        super().__init__('feasibility_test_node')
        
        # Publisher: /ur10skku/cmdMotion (9차원 데이터 송신)
        self.cmd_motion_pub = self.create_publisher(Float64MultiArray, '/ur10skku/cmdMotion', 10)
        
        # Subscribers
        self.current_p_sub = self.create_subscription(
            Float64MultiArray, '/ur10skku/currentP', self.current_p_callback, 10)
            
        # [수정 2] ftdata의 구독 타입을 WrenchStamped로 변경
        self.ft_data_sub = self.create_subscription(
            WrenchStamped, '/ur10skku/ftdata', self.ft_data_callback, 10)
            
        self.current_pos = None  # [x, y, z, wx, wy, wz]
        self.current_ft = None   # [fx, fy, fz, mx, my, mz]
        
        # [데이터 고정] 하위 6개 값 세팅
        self.fixed_wx = 0.0
        self.fixed_wy = 0.0
        self.fixed_wz = 1.57
        self.fixed_fx = 0.0
        self.fixed_fy = 0.0
        self.fixed_fz = 0.0

        # 제어 주기 및 하강 설정
        self.timer_period = 0.01  # 100Hz (0.01초)
        self.z_step_mm = 0.5      # Z축 하강 속도 (1틱당 0.5mm 하강)
        
        self.timer = self.create_timer(self.timer_period, self.control_loop)
        self.is_contacted = False
        
        self.get_logger().info("피지빌리티 테스트 노드 실행: 수직 강하 대기 중...")

    def current_p_callback(self, msg):
        self.current_pos = msg.data

    # [수정 3] WrenchStamped 구조에 맞게 데이터 파싱 로직 변경
    def ft_data_callback(self, msg):
        fx = msg.wrench.force.x
        fy = msg.wrench.force.y
        fz = msg.wrench.force.z
        mx = msg.wrench.torque.x
        my = msg.wrench.torque.y
        mz = msg.wrench.torque.z
        self.current_ft = [fx, fy, fz, mx, my, mz]

    def contact_point_detection_algorithm(self, pos, ft):
        self.get_logger().info("접촉 유지 중: 접촉점 탐지 알고리즘 구동...", throttle_duration_sec=1.0)
        x, y, z = pos[0], pos[1], pos[2]
        return x, y, z 

    def control_loop(self):
        # 센서 데이터가 모두 수신될 때까지 대기
        if self.current_pos is None or self.current_ft is None:
            return

        x, y, z, _, _, _ = self.current_pos
        _, _, fz, _, _, _ = self.current_ft

        cmd_msg = Float64MultiArray()

        # 접촉 임계값 설정
        contact_threshold = 2.0 
        
        if abs(fz) > contact_threshold:
            if not self.is_contacted:
                self.get_logger().info(f"표면 접촉 감지! (Fz: {fz:.2f}N). 탐지 알고리즘으로 전환합니다.")
                self.is_contacted = True
            
            target_x, target_y, target_z = self.contact_point_detection_algorithm(self.current_pos, self.current_ft)
            
            cmd_msg.data = [
                target_x, target_y, target_z, 
                self.fixed_wx, self.fixed_wy, self.fixed_wz, 
                self.fixed_fx, self.fixed_fy, self.fixed_fz
            ]
        else:
            self.is_contacted = False
            target_z = z - self.z_step_mm

            cmd_msg.data = [
                x, y, target_z, 
                self.fixed_wx, self.fixed_wy, self.fixed_wz, 
                self.fixed_fx, self.fixed_fy, self.fixed_fz
            ]

        self.cmd_motion_pub.publish(cmd_msg)

def main(args=None):
    rclpy.init(args=args)
    node = FeasibilityTestNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("노드가 종료되었습니다.")
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/contact_pkg/src/feasibility_test_node.cpp
#include <array>
#include <cmath>
#include <memory>
#include <vector>

<<<<<<< HEAD

=======
>>>>>>> origin/main
#include "geometry_msgs/msg/wrench_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"

class FeasibilityTestNode : public rclcpp::Node
{
public:
  FeasibilityTestNode()
  : Node("feasibility_test_node")
  {
    cmd_motion_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>("/ur10skku/cmdMotion", 10);

    current_p_sub_ = create_subscription<std_msgs::msg::Float64MultiArray>(
      "/ur10skku/currentP", 10,
      [this](const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
        if (msg->data.size() >= 6) {
          for (size_t i = 0; i < current_pos_.size(); ++i) {
            current_pos_[i] = msg->data[i];
          }
          has_current_pos_ = true;
        } else {
          RCLCPP_WARN_THROTTLE(
            get_logger(), *get_clock(), 2000,
            "currentP 데이터 길이가 부족합니다. expected >= 6, got %zu", msg->data.size());
        }
      });

    ft_data_sub_ = create_subscription<geometry_msgs::msg::WrenchStamped>(
      "/ur10skku/ftdata", 10,
      [this](const geometry_msgs::msg::WrenchStamped::SharedPtr msg) {
        current_ft_ = {
          msg->wrench.force.x,
          msg->wrench.force.y,
          msg->wrench.force.z,
          msg->wrench.torque.x,
          msg->wrench.torque.y,
          msg->wrench.torque.z};
        has_current_ft_ = true;
      });

    timer_ = create_wall_timer(
      std::chrono::milliseconds(10),
      std::bind(&FeasibilityTestNode::controlLoop, this));

    RCLCPP_INFO(get_logger(), "피지빌리티 테스트 노드 실행: 수직 강하 대기 중...");
  }

private:
  std::array<double, 3> contactPointDetectionAlgorithm(
    const std::array<double, 6> & pos,
    const std::array<double, 6> & /*ft*/)
  {
    RCLCPP_INFO_THROTTLE(
      get_logger(), *get_clock(), 1000,
      "접촉 유지 중: 접촉점 탐지 알고리즘 구동...");
    return {pos[0], pos[1], pos[2]};
  }

  void controlLoop()
  {
    if (!has_current_pos_ || !has_current_ft_) {
      return;
    }

    const double x = current_pos_[0];
    const double y = current_pos_[1];
    const double z = current_pos_[2];
    const double fz = current_ft_[2];

    std_msgs::msg::Float64MultiArray cmd_msg;

    if (std::abs(fz) > contact_threshold_) {
      if (!is_contacted_) {
        RCLCPP_INFO(
          get_logger(),
          "표면 접촉 감지! (Fz: %.2fN). 탐지 알고리즘으로 전환합니다.", fz);
        is_contacted_ = true;
      }

      const auto target = contactPointDetectionAlgorithm(current_pos_, current_ft_);
      cmd_msg.data = {
        target[0], target[1], target[2],
        fixed_wx_, fixed_wy_, fixed_wz_,
        fixed_fx_, fixed_fy_, fixed_fz_};
    } else {
      is_contacted_ = false;
      const double target_z = z - z_step_mm_;
      cmd_msg.data = {
        x, y, target_z,
        fixed_wx_, fixed_wy_, fixed_wz_,
        fixed_fx_, fixed_fy_, fixed_fz_};
    }

    cmd_motion_pub_->publish(cmd_msg);
  }

  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr cmd_motion_pub_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr current_p_sub_;
  rclcpp::Subscription<geometry_msgs::msg::WrenchStamped>::SharedPtr ft_data_sub_;
  rclcpp::TimerBase::SharedPtr timer_;

  std::array<double, 6> current_pos_{};
  std::array<double, 6> current_ft_{};
  bool has_current_pos_{false};
  bool has_current_ft_{false};
  bool is_contacted_{false};

  const double fixed_wx_{0.0};
  const double fixed_wy_{0.0};
  const double fixed_wz_{1.57};
  const double fixed_fx_{0.0};
  const double fixed_fy_{0.0};
  const double fixed_fz_{0.0};
  const double z_step_mm_{0.5};
  const double contact_threshold_{2.0};
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<FeasibilityTestNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
